Enhancements for hashed bin delegation

joshuagl opened the pull request:
**Fixes issue #**: #994, #995, #1003, #1004

**Description of the changes being introduced by the pull request**:

__Note__: This is the first PR addressing changes required by the PEP 458 implementation.

* Make tests around hashed bins simpler and more robust
* Add various helper functions for common functionality around hashed bins:
  * `_get_hash()` to hash a filepath string
  * `_create_bin_name()` to create the string name of a delegated hash bin (mostly to consolidate the instances where we do non-trivial string formatting)
  * `_find_bin_for_hash()` to calculate the bin a hashed file string would be allocated to
* Refactor `add_target_to_bin()` and `remove_target_from_bin()` to remove the dispatcher and simplify the code as well as allowing `add_target_to_bin()` to take a custom argument, the same as `add_target()`
* Refactor `delegate_hashed_bins()` to simplify the code
* Ensure file paths are hashed as they appear in the targets metadata
* Enable passing a fileinfo to `add_target()` and `add_target_to_bin()`
* Add `use_existing_fileinfo` argument to `write()` and `writeall()` to allow writing metadata without access to targets files

**Please verify and check that the pull request fulfils the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

cc @woodruffw

joshuagl committed:
Add _get_hash() helper function to repository_tool

This helper will generate a hash of the passed target_filepath using the
algorithm defined by the repository_tool's HASH_FUNCTION variable.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Add _find_bin_for_hash() helper to repository_tool

Add a helper function to determine the name of a bin that a hashed
targetfile will be delegated to.

Based sketches by Lukas Puehringer in issues #994 & #995

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Simplify delegated hash bin tests

As we are adding and removing items from the hashed bins and checking
for their presence/absence it's simplest if we being with the hashed
bins initially empty.

If we pass a list of targets when we call delgate_hashed_bins() the
delegated roles have an initial set of targets delegated to them,
which complicates testing of adding then removing a target to a
delegated bin.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Refactor add_target_to_bin & remove_target_from_bin

Vastly simplify the implementation, using the _get_hash() and
_find_bin_for_hash() helpers added in earlier commits.

Furthermore, enable passing of the custom parameter to
add_target_to_bin() to better match add_target()

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Test to ensure delegated bin names match prefixes

Add test to ensure delegated bin names are consistent with the hash
prefixes that are delegated to the role.

This is an implicit assumption of the current implementation, the
testing of which will enable us to modify the code with greater
confidence.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Refactor delegate_hashed_bins

Simplify the delegate_hashed_bins function based on the semi-pseudo-code
by Lukas Puehringer in #995.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Hash target file path's as they appear in metadata

In delegate_to_hashed_bins() duplicate the code which determines the
target's file path for writing to the metadata in repository_lib's
generate_targets_metadata to ensure that the target path hashed by
delegate_hashed_bins() matches the target path in the metadata file.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Do not check target file exists in add_target[s]()

The file isn't strictly needed on-disk at the time add_target() and
add_targets() are called and this duplicates the check for the file's
presence in write[_all]()

By removing this check we allow extra versatility in adding targets.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Enable passing a fileinfo to add_target[_to_bin]()

Add an additional optional parameter to add_target() and
add_target_to_bin() which is a fileinfo object matching
tuf.formats.FILEINFO_OBJECT

This parameter and the custom parameter are mutually exclusive and
thus cannot be passed at the same time.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Fix replacing targets in add_targets()

Match the pattern in add_target() where if the filepath already exists
in roleinfo['paths'] it is updated to replace the existing entry with
the new fileinfo.

Signed-off-by: Joshua Lock <jlock@vmware.com>

woodruffw commented on tuf/repository_tool.py:
```diff
@@ -2506,89 +2502,47 @@ def delegate_hashed_bins(self, list_of_targets, keys_of_hashed_bins,
     logger.info(repr(len(list_of_targets)) + ' total targets.')
```
Nitpick: Could these separate `info` calls be merged into a single message? That would eliminate the chance of them being spread when integrated into a logging system with multiple parallel sources.

joshuagl replied:
Good idea, I've pushed that in a follow-on patch ed93fcc

woodruffw commented on tuf/repository_tool.py:
```diff
@@ -1953,31 +1961,28 @@ def add_target(self, filepath, custom=None):
+    # targets directory.
+    targets_directory_length = len(self._targets_directory) + 1
+    roleinfo = tuf.roledb.get_roleinfo(self._rolename, self._repository_name)
+    relative_path = filepath[targets_directory_length:].replace('\\', '/')
```
Nit (can probably be ignored in this PR): It might be good to add (or use) a path canonicalization method instead of calling `replace('\\', '/')` throughout the codebase.

joshuagl replied:
Agreed, this would be good cleanup.

@lukpueh any thoughts on whether this belongs in securesystemslib or in tuf? Perhaps as part of a new utils module?

lukpueh replied:
Absolutely! https://github.com/theupdateframework/tuf/pull/1008#pullrequestreview-382660018 has some thoughts about path normalization. The TLDR is, if the caller passes a *bad* path,
- either fail,
- or warn and normalize

In both cases this should happen once at the boundary where a path is added to the system, and thus remove the need for many `replace('\\', '/')` throughout the code.

joshuagl committed:
Allow writing metadata without access to targets

Previously, at the time of writing targets metadata, we accessed all
targets files and ensured they exist on disk before generating a
filinfo dict containing hashes of the target file.

This change enables targets metadata to be generated using the
fileinfo data that is already stored in the roledb.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Only call logger once for hashed bin info

Merge the logger calls reporting information about the hashed bin
delegations into a single logger.info() call to ensure the messages
will be grouped together even when integrated into a logging system
with multiple parallel sources.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
I've pushed a couple of extra patches to this branch, one to address a review nit from @woodruffw  and another to address #1004

joshuagl committed:
Add test for repo without access to target files

Test the newly added functionality to:
* add a target to the repository without access to the target file on disk
* write targets metadata without access to target files on disk, by using
  the existing fileinfo data from the roledb

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh commented on tests/test_repository_tool.py:
```diff
@@ -1288,6 +1288,27 @@ def test_delegate_hashed_bins(self):
+      roleinfo = tuf.roledb.get_roleinfo(self.targets_object.rolename,
+          'test_repository')
+      for delegated_role in roleinfo['delegations']['roles']:
+        if len(delegated_role['path_hash_prefixes']) > 0:
```
Minor nit: This is a great check, it would a shame if it were silently omitted because of a hiccup in the roledb that makes so that we only get empty `delegated_role['path_hash_prefixes']` lists. Maybe worth a tiny TODO: comment?

joshuagl replied:
Great point @lukpueh. 

When looking at the patch just now I realised that we only ever call this function when we know we have at least one delegation with `path_hash_prefixes`, so I'm going to add a flag which is set when we find a role with `len(delegated_role['path_hash_prefixes']) > 0` and assert that the flag is set at the end of the `for` loop. That was we should catch a hiccup in the roledb like you described.

joshuagl replied:
Added the change I mentioned above as a follow-on patch f362544

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2824,6 +2824,59 @@ def _get_hash(target_filepath):
+  prefix_len = len("{:x}".format(number_of_bins - 1))
+  prefix_count = 16 ** prefix_len
+
+  if prefix_count % number_of_bins != 0:
```
Minor nit: Maybe it's worth a comment that `x % y != 0` does not guarantee  that `y` is not a power of two for any `x` and `y`, but due to the relationship between `number_of_bins` and `prefix_count` it is true for them.

joshuagl replied:
That's a good point, I've updated this patch to include a comment

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2669,107 +2693,20 @@ def remove_target_from_bin(self, target_filepath):
 
-    else:
+    # Ensure the Targets object has delegated to hashed bins
+    if not self._delegated_roles.get(bin_name, None):
```
Would you mind also fixing the error message to what you use above? 
```
      raise securesystemslib.exceptions.Error(self.rolename + ' does not have'
          ' a delegated role ' + bin_name)
```

joshuagl replied:
Good catch, thanks. I've updated the patch to make this consistent.

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2486,14 +2486,13 @@ def delegate_hashed_bins(self, list_of_targets, keys_of_hashed_bins,
-    prefix_length =  len(hex(number_of_bins - 1)[2:])
+    prefix_length = len("{:x}".format(number_of_bins - 1))
     total_hash_prefixes = 16 ** prefix_length
+    bin_size = total_hash_prefixes // number_of_bins
```
Minor nit: Maybe above three lines together with the "power of two" check below are worth a helper, to be used here and in `_find_bin_for_hash()`.

joshuagl replied:
The rest of the code uses all three of the variables (though one only in a comment iirc) and I'm not sure about adding a function that returns multiple values, so I've left this suggestion for now.

trishankatdatadog replied:
There should also be a comment that this particular LoC is _rescaling_ the index. I didn't really understand what was going on until I looked above to see what `number_of_bins` is.

lukpueh replied:
I'm fine with leaving it as it is, although it is the same code  (with different names and order) in both cases.

`delegate_hashed_bins`:
https://github.com/theupdateframework/tuf/blob/5ea1f6398a27047a5a299c42b7428c6248d45e0e/tuf/repository_tool.py#L2504-L2513

`_find_bin_for_hash`:
https://github.com/theupdateframework/tuf/blob/5ea1f6398a27047a5a299c42b7428c6248d45e0e/tuf/repository_tool.py#L2768-L2777

And I personally see no problems with idioms like 

```python
pefix_length, prefix_count, bin_size = _get_bin_numbers(number_of_bins)
```

Also, and this was part of my motivation, we'd only need one place to properly comment on this math-magic (thanks for pointing that out, @trishankatdatadog).

Btw., @trishankatdatadog, I wouldn't say that this particular LoC rescales the index. Did you maybe mean this one...?
https://github.com/theupdateframework/tuf/blob/5ea1f6398a27047a5a299c42b7428c6248d45e0e/tuf/repository_tool.py#L2547

Yes, the general idea here is that we scale `0..total_hash_prefixes` to `0..number_of_bins`. But I wouldn't say that the particular LoC pointed out by you does this.

joshuagl replied:
Convincing points. I will factor out this into a helper now.

joshuagl replied:
Helper added, with lots of comments, in ec68416

trishankatdatadog replied:
@lukpueh Yes, I meant that line, the one with the integer division

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2506,84 +2505,41 @@ def delegate_hashed_bins(self, list_of_targets, keys_of_hashed_bins,
-    for bin_index in six.moves.xrange(total_hash_prefixes):
-      target_paths_in_bin[bin_index] = []
+      role = {"name": name, "target_paths": [],
+          'target_hash_prefixes': target_hash_prefixes}
```
Minor nit: I think dict literals are easier to read when each item is on a separate line.

joshuagl replied:
I've squashed in a change to make each item on a separate line _and_ use consistent quotation marks.

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2551,10 +2566,17 @@ def add_target_to_bin(self, target_filepath, number_of_bins, custom=None):
+      fileinfo:
+        An optional fileinfo object, conforming to tuf.formats.FILEINFO_SCHEMA,
+        providing full information about the file.
+
```
Given that we said that the only reason to have `custom` and `fileinfo`  is to support backwards compatibility (as the former can be passed within the latter) and `add_target_to_bin` didn't support this before, I think it is safe to not add `custom` here.

lukpueh replied:
I understand that my suggestion comes at the cost of consistency between `add_target` and `add_target_to_bin`. I am fine either way.

joshuagl replied:
I agree. As we plan to deprecate the `custom` parameter it doesn't make sense to _add_ it to `add_target_to_bin`. I've added a follow-on patch in 5ea1f63, though maybe I should have squashed it.

lukpueh commented on tuf/repository_lib.py:
```diff
@@ -1284,41 +1304,22 @@ def generate_targets_metadata(targets_directory, target_files, version,
-    # Note: join() discards 'targets_directory' if 'target' contains a leading
-    # path separator (i.e., is treated as an absolute path).
-    target_path = os.path.join(targets_directory, target.lstrip(os.sep))
+      filedict[target.replace('\\', '/').lstrip('/')] = fileinfo
```
Unrelated: I think this should have happened before (see https://github.com/theupdateframework/tuf/pull/1007#discussion_r399222066)

joshuagl replied:
Hopefully it's OK to remove this as we consolidate path handling under #1008, or as a follow on to those changes?

woodruffw replied:
Under #1008 or a follow on is okay for me!

lukpueh commented on tuf/repository_lib.py:
```diff
@@ -1284,41 +1304,22 @@ def generate_targets_metadata(targets_directory, target_files, version,
-    # (e.g., 'targets/more_targets/somefile.txt' -> 'more_targets/somefile.txt')
-    relative_targetpath = target
+      # Ensure all fileinfo entries in target_files have a non-empty hashes dict
+      if not fileinfo.get('hashes', None):
```
Important: This should also fail if 'length' is missing.

joshuagl replied:
Great catch, thank you. I've squashed a change to implement this into the patch.

lukpueh commented on tuf/repository_lib.py:
```diff
@@ -1352,6 +1353,58 @@ def generate_targets_metadata(targets_directory, target_files, version,
+
+    # Ensure all target files listed in 'target_files' exist.  If just one of
+    # these files does not exist, raise an exception.
+    if not os.path.exists(target_path):
```
Absolutely unrelated: This is the first of three `path.exists`/ `path.isfile` checks in the stack before we open that file (`get_metadata_fileinfo` has one and `securesystemslib.util.get_file_details` has one too). As good Pythonistas, we should ask for forgiveness instead.

joshuagl replied:
We should absolutely move towards the Python idiom of [EAFP](https://docs.python.org/3/glossary.html#term-eafp). I think we can do that in future changes as we work to address #1009

lukpueh reviewed (changes requested):
This is impeccable work, @joshuagl. I only have one blocking comment, which is amazing for a PR of that size. That comment starts with "Important:". All the other comments are just sugar, so that you see I actually reviewed this. :P 

Many many thanks!

joshuagl committed:
Test at least one delegated bin has path_hash_prefixes

When testing delegate_hashed_bins to ensure that hash_path_prefixes
map to the generated name of the bin, also check to ensure that at least
one of the delegations contains one or more path_hash_prefixes.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Remove custom parameter from add_target_to_bin()

We intend to deprecate the custom parameter of add_target() in favour
of using the fileinfo parameter with the custom value populated,
therefore it does not make sense to _add_ the custom parameter to
add_target_to_bin()

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
Thank you for the thorough review @lukpueh . I've addressed most of your comments, some in squashed commits and some in follow-on patches.

I have replied to all of your comments so that you can see where I've made changes.

joshuagl committed:
Factor out bin numbers calculations into a helper

Factor out code to calculate the prefix length, total number of prefixes
and number of prefixes to be allocated to each bin into a helper
function _get_bin_numbers() so that we a) do not duplicate logic,
b) can document the logic in one place and c) ensure the logic is
consistent.

Signed-off-by: Joshua Lock <jlock@vmware.com>

woodruffw commented on tests/test_repository_tool.py:
```diff
@@ -369,6 +369,108 @@ def test_writeall(self):
     self.assertRaises(securesystemslib.exceptions.FormatError, repository.writeall, 3)
 
 
+  def test_writeall_no_files(self):
```
Just wanted to give kudos for adding a test case that closely mirrors the current work on Warehouse ❤️

lukpueh commented on tuf/repository_lib.py:
```diff
@@ -1284,43 +1304,26 @@ def generate_targets_metadata(targets_directory, target_files, version,
-    custom_data = None
-    if len(custom):
-      custom_data = custom
+      if fileinfo.get('length', -1) < 0:
```
Cool way of checking this. 👍

joshuagl replied:
Feels a bit C-like, but I couldn't think of a cleaner way to do it.

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2626,12 +2586,24 @@ def add_target_to_bin(self, target_filepath):
+          ' a delegated role ' + bin_name)
 
-    return self._locate_and_update_target_in_bin(target_filepath, 'add_target')
+    self._delegated_roles[bin_name].add_target(target_filepath, fileinfo)
```
```suggestion
    self._delegated_roles[bin_name].add_target(target_filepath, fileinfo=fileinfo)
```
... otherwise `fileinfo` is passed as `custom` to `add_target`.

It would be great to have a test for `add_target_to_bin` that adds a `fileinfo`.

joshuagl replied:
Great catch Lukas. I've squashed this fix and added a test in 62e4364

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2669,135 +2645,150 @@ def remove_target_from_bin(self, target_filepath):
+    # Ensure the Targets object has delegated to hashed bins
+    if not self._delegated_roles.get(bin_name, None):
+      raise securesystemslib.exceptions.Error(self.rolename + ' does not have'
+          ' a delegated roles ' + bin_name)
```
```suggestion
          ' a delegated role ' + bin_name)
```

lukpueh commented on tuf/repository_tool.py:
```diff
@@ -2669,135 +2645,150 @@ def remove_target_from_bin(self, target_filepath):
+
+  # For simplicity, ensure that 'prefix_count' (16 ^ n) can be evenly
+  # distributed over 'number_of_bins' (must be 2 ^ n).  Each bin will contain
+  # (prefix_count / number_of_bins) hash prefixes.
```
Ping https://github.com/theupdateframework/tuf/pull/1007#discussion_r399269492

joshuagl replied:
Thanks for catching this. My original comment was lost when factoring out into the helper. I've updated that patch to add a comment as suggested.

lukpueh reviewed (changes requested):
Thanks for the updates! ❤️ Here are some more comments.

joshuagl committed:
Add a test for add_target_to_bin that adds a fileinfo

Add some additional checks to test_add_target_to_bin to ensure the code
to add a target passing a fileinfo is tested.

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh reviewed (approved):
🚀
